### ISSUE TITLE
Enabling GPU mixed precision algorithms 

### ISSUE THREAD

**Issue by jvitordeoliveira96** [pytorch/pytorch#132940]
### 🚀 The feature, motivation and pitch

PyTorch does not seem to provide wrappers to mixed precision algorithms in e.g MAGMA (dshpov, shportf) and cuSolver (https://docs.nvidia.com/cuda/cusolver/index.html#cusolverirsrefinement-t). Does anyone know if that is really the case? 
It would be interesting to have access to these routines since these may potentially improve performance of linear system solving by a factor 4.7 using GPUs with negligible changes in the accuracy (https://epubs.siam.org/doi/10.1137/19M1298263). 

I created two rough prototypes for the changes that would be needed to make those mixed precision algorithms available. It seems to be a simple modification. My prototype works by changing the behavior of functions cholesky_solve (for using the dshpov MAGMA routine) and linalg.cholesky (for using the shportf MAGMA routine). Due to my lack of deep knowledge of PyTorch, I was unable to go further.  It would be really awesome if someone could work on top of these prototypes to make a proper pull request.

One package that could benefit from these additions would be GPyTorch, since it executes Cholesky factorizations repeatedly during its training and prediction procedures. Since GPyTorch only uses solvers from PyTorch, it seems more reasonable to make these changes here.

Let me know if I can give more information!

Links for the prototypes: https://github.com/pytorch/pytorch/commit/a88f53c786d437d1cee49532aa1d240eda7c855e
https://github.com/pytorch/pytorch/commit/8c7194acff4924530dc63ca3e0b70fb7af8cc182

**Comment by lezcano** [comment]
I think the idea is quite interesting. I see the implementation uses `magma_dshposv_native` but the docs just document [`magma_dshposv_gpu_expert`](https://icl.utk.edu/projectsfiles/magma/doxygen/group__magma__posv.html#ga1a5390b78aee3f77d5c1c08d8bbc965e). Do you know what's the expected API to use?

Note that this approach would be quite limited in its uses in as it would just apply to systems of one matrix. Neither magma nor cusolver seem expose a batched version of these. Even more, in the magma case it just exposed it for the dtype double.

The cusolver mixed precision methods for `solve`  and `lstsq` expose a pretty large API surface though, and it's not clear to me how to efficiently dispatch to the different methods here without exposing all this API via an extra parameter to `linalg.solve` of the form `interal_precision: Optional[torch.dtype]`. 

Do you know what is the intersection and difference of the MAGMA / cuSolver APIs?

It would also be good to get some benchmarks to see what are the speed-ups of these methods on matrices of size 256 / 512 / 1024 / 2048 cross product some rhs sizes. Could you look into this? Note that in your prototypes you are making unnecessary copies of the inputs, while you already have the input available and you could just pass that one in. Making an unnecessary copy will surely negate any perf gains coming from using these methods. To benchmark, you can use the [pytorch benchmark timers](https://pytorch.org/tutorials/recipes/recipes/benchmark.html).

**Comment by lezcano** [comment]
@IvanYashchuk any thoughts?

**Comment by KeAWang** [comment]
Any updates on this? I'm looking for a CUDA accelerated mixed precision linear system solver for spd matrices (like what gpytorch needs)
